### PR TITLE
Add calendar links

### PR DESCRIPTION
**Pull request by brianvans**
Drops off a calendar link at the end of each description

### CODE REVIEW

File: index.rb [hunk 1]
```diff
@@ -11,7 +12,33 @@
...
+    Base64.urlsafe_encode64(
+      [
+        ['BEGIN', 'VCALENDAR'],
+        ['VERSION', '2.0'],
+        ['BEGIN', 'VEVENT'],
+        ['URL', link],
+        ['DTSTART', ical_time(time)],
+        ['DTEND', ical_time(time + (3600 * 3))],
+        ['SUMMARY', title],
+        ['DESCRIPTION', description&.gsub("\n", '\\n')],
```
Comment (davishmcclurg) [inline index.rb:27]: Will this have html in it? Does that matter?

File: index.rb [hunk 2]
```diff
@@ -243,6 +270,8 @@ def normalize(text)
                 <strong><a href="<%= h(show.link) %>"><%= normalize(show.title) %></a></strong>
                 <br>
                 <%= normalize(show.description) %>
+                <br>
+                <a href="<%= show.ical_link %>">iCal</a>
```
Comment (davishmcclurg) [inline index.rb:274]: Maybe we should make the date/time text link to this instead?

**Review by davishmcclurg — APPROVED** [review]

File: index.rb (hunk 1 above)
Comment (brianvans) [reply index.rb:27]: I don't think the DESCRIPTION field supports HTML, so if some sneaks in, I don't think it matters.

There might be other characters we need to escape, according to the spec: https://icalendar.org/iCalendar-RFC-5545/3-3-11-text.html

File: index.rb (hunk 2 above)
Comment (brianvans) [reply index.rb:274]: I'm all for that

File: index.rb (hunk 1 above)
Comment (davishmcclurg) [reply index.rb:27]: > I don't think the DESCRIPTION field supports HTML, so if some sneaks in, I don't think it matters.

It looks like it's breaking for Bottom of the Hill links: https://github.com/davishmcclurg/shows/pull/14

File: index.rb (hunk 2 above)
Comment (davishmcclurg) [reply index.rb:274]: Also here: https://github.com/davishmcclurg/shows/pull/14

File: index.rb [hunk 3]
```diff
@@ -11,7 +12,33 @@
...
+        ['END', 'VCALENDAR']
+      ].map { |e| e.join(':') }.join("\n")
+    )
+  end
+
+  private
+
+  # https://icalendar.org/iCalendar-RFC-5545/3-3-5-date-time.html
+  def ical_time(time)
+    time.utc.strftime('%Y%m%dT%H%M%SZ')
```
Comment (davishmcclurg) [inline index.rb:39]: `utc` messed with our time objects, which is surprising: https://github.com/davishmcclurg/shows/commit/e61048b4493e58b4da4361ef211c47b0e9563380

File: index.rb [hunk 4]
```diff
@@ -11,7 +12,33 @@
   gem 'nokogiri'
 end
 
-Show = Struct.new(:time, :link, :title, :description, :venue, :keyword_init => true)
+Show = Struct.new(:time, :link, :title, :description, :venue, :keyword_init => true) do
+  def ical_link
+    'data:text/calendar;base64,' +
+    Base64.urlsafe_encode64(
```
Comment (davishmcclurg) [inline index.rb:18]: Had to change this to `strict_encode64`: https://github.com/davishmcclurg/shows/commit/29bcd5b3b4d72e9226189793728378b95ba39fcd

**Pull request merged**